### ISSUE TITLE
Sparse matrix index dtype on Windows differs for CSR/CSC

### ISSUE THREAD

**Issue by c-f-h** [scipy/scipy#11641]
I'm testing this on a 64 bit Anaconda installation on Windows 10.

Creating a diag sparse matrix and converting it to CSR or CSC results in two different buffer dtypes for the `indices` array of the resulting sparse matrix:

#### Reproducing code example:
```
>>> memoryview(scipy.sparse.eye(10).tocsr().indices).format
'i'
>>> memoryview(scipy.sparse.eye(10).tocsc().indices).format
'l'
```

Although both 'i' and 'l' refer to a 32 bit signed integer on Windows, this makes it trickier to check for the correct buffer dtype in externally compiled matrix manipulation routines.

#### Scipy/Numpy/Python version information:
```
1.4.1 1.18.1 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)
```

**Comment by perimosocordiae** [comment]
Interesting, I can't reproduce on an Ubuntu machine on master (both versions produce 'i' for me).

This probably arises from the fact that DIA to CSR conversions route through COO format, whereas DIA to CSC has a specialized implementation.

Can someone with a Win10 dev machine reproduce on master?

**Comment by c-f-h** [comment]
> Interesting, I can't reproduce on an Ubuntu machine on master (both versions produce 'i' for me).

Yes. Sorry, I forgot to mention that in Linux, both lines result in 'i'. This seems to be Windows-specific.

**Comment by ewmoore** [comment]
Yes, this reproduces on windows.
```
In [1]: import numpy as np

In [2]: from scipy import sparse

In [3]: adia = sparse.eye(10)

In [4]: acsr = adia.tocsr()

In [5]: acsr.indices.dtype.char
Out[5]: 'i'

In [6]: acsr.indptr.dtype.char
Out[6]: 'i'

In [7]: acsc = adia.tocsc()

In [8]: acsc.indices.dtype.char
Out[8]: 'l'

In [9]: acsc.indptr.dtype.char
Out[9]: 'i'

In [10]: acsc.indices.dtype
Out[10]: dtype('int32')

In [11]: acsc.indptr.dtype
Out[11]: dtype('int32')
```
I think the issue is that the `astype` call [here](https://github.com/scipy/scipy/blob/master/scipy/sparse/dia.py#L352) doesn't actually make sure the dtype is what we want.  For instance:
```
In [15]: q = np.arange(10)

In [16]: q.dtype.char
Out[16]: 'l'

In [17]: q.astype('i', copy=False).dtype.char
Out[17]: 'l'
```
If we drop the `copy=False` it will work, 
```
In [19]: q.astype('i').dtype.char
Out[19]: 'i'
```

What are the guarantees we provide on the dtype of `indices` and `indptr`.  I wonder if the constructors should be enforcing something even when called with existing `indices`, `indptr` and a `data` arrays?

**Comment by perimosocordiae** [comment]
That astype behavior is quite surprising to me. Is that a numpy bug, or am I not understanding properly?

**Comment by ewmoore** [comment]
It’s not a numpy bug.  `astype` uses [`PyArray_EquivTypes`](https://github.com/numpy/numpy/blob/fed1fb49bc50fd2a44238b8250d13314a127ddcc/numpy/core/src/multiarray/multiarraymodule.c#L1464) which says two dtypes are the same if they have the same size, byte order, and basic kind (e.g. signed integers).  Since as far as numpy dtypes go they are just different names for the same type if these things are all true. 

Perhaps @c-f-h should just call `PyArray_EquivTypes` when trying to consume these arrays.  Our guarantee could just be that the types will always be equivalent even if the type codes are different.  I guess this is more difficult if consuming the arrays as memoryviews and not ndarrays, since all you have in that case is a type code and not a dtype.

**Comment by c-f-h** [comment]
I hit upon this while writing wrappers for C++ sparse matrix routines using pybind11. What I have there is a format string (more specifically, see the `buffer_info` type described at https://pybind11.readthedocs.io/en/master/advanced/pycpp/numpy.html).

This is certainly something that can be worked around, but the resulting code feels somewhat kludgy. If there is no simple fix, so be it.
